File: cpp/020.ValidParentheses/ValidParentheses.cpp
// O(n)
// Runtime Beats 100.00%

#include <cassert>
#include <iostream>
#include <stack>
#include <string>
#include <unordered_map>

class Solution {
public:
    bool isValid(std::string s) {
        std::unordered_map<char, char> map{{'(', ')'}, {'{', '}'}, {'[', ']'}};

        std::stack<char> stack;
        for (const auto& c : s) {
            if (map.find(c) != map.end()) {
                stack.push(c);
            } else {
                if (stack.empty()) {
                    return false;
                }
                auto top = stack.top();
                if (map[top] != c) {
                    return false;
                }

                stack.pop();
            }
        }

        return stack.empty();
    }
};

void test1() {
    Solution s;
    assert(s.isValid("()") == true);
}

void test2() {
    Solution s;
    assert(s.isValid("()[]{}") == true);
}

void test3() {
    Solution s;
    assert(s.isValid("(]") == false);
}

void test4() {
    Solution s;
    assert(s.isValid("([)]") == false);
}

void test5() {
    Solution s;
    assert(s.isValid("{[]}") == true);
}

int main() {
    test1();
    test2();
    test3();
    test4();
    test5();
    return 0;
}

File: python3/020.ValidParentheses/ValidParentheses.py
# O(n)
# Runtime Beats 100.00%


class Solution:

    def isValid(self, s: str) -> bool:
        stack = []
        paren_map = {')': '(', ']': '[', '}': '{'}
        for c in s:
            if c not in paren_map:
                stack.append(c)
            elif not stack or paren_map[c] != stack.pop():
                return False
        return not stack


def test1():
    s = Solution()
    assert s.isValid("()") == True


def test2():
    s = Solution()
    assert s.isValid("()[]{}") == True


def test3():
    s = Solution()
    assert s.isValid("(]") == False


def test4():
    s = Solution()
    assert s.isValid("([)]") == False


def test5():
    s = Solution()
    assert s.isValid("{[]}") == True


test1()
test2()
test3()
test4()
test5()
